pkg/bucketaccess/bucketaccess_controller.go

```diff
@@ -0,0 +1,248 @@
+type BucketAccessListener struct {
+	provisionerClient cosi.ProvisionerClient
+	provisionerName   string
+
+	kubeClient   kube.Interface
+	bucketClient buckets.Interface
+	kubeVersion  *utilversion.Version
+
+	namespace string
+}
+
+// NewBucketAccessListener returns a resource handler for BucketAccess objects
+func NewBucketAccessListener(provisionerName string, client cosi.ProvisionerClient) (*BucketAccessListener, error) {
+	ns := os.Getenv("POD_NAMESPACE")
+	if ns == "" {
+		return nil, errors.New("POD_NAMESPACE env var cannot be empty")
+	}
+
+	return &BucketAccessListener{
+		provisionerName:   provisionerName,
+		provisionerClient: client,
+		namespace:         ns,
+	}, nil
+}
+
+// Add attempts to create a bucket for a given bucket. This function must be idempotent
+// Return values
+//    nil - BucketAccess successfully granted
+//    non-nil err - Internal error                                [requeue'd with exponential backoff]
+func (bal *BucketAccessListener) Add(ctx context.Context, bucketAccess *v1alpha1.BucketAccess) error {
+	bucketName := bucketAccess.Spec.BucketName
+
+	klog.V(3).InfoS("Add BucketAccess",
+		"name", bucketAccess.Name,
+		"bucket", bucketName,
+	)
+
+	if bucketAccess.Spec.MintedSecretName != "" {
+		klog.V(5).InfoS("AccessAlreadyGranted",
+			"bucketAccess", bucketAccess.Name,
+			"bucket", bucketName,
+		)
+		return nil
+	}
+
+	bucket, err := bal.Buckets().Get(ctx, bucketName, metav1.GetOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Failed to fetch bucket", "bucket", bucketName)
+		return errors.Wrap(err, "Failed to fetch bucket")
+	}
+
+	if !strings.EqualFold(bucket.Spec.Provisioner, bal.provisionerName) {
+		klog.V(5).InfoS("Skipping bucketaccess for provisiner",
+			"bucketAccess", bucketAccess.Name,
+			"provisioner", bucket.Spec.Provisioner,
+		)
+		return nil
+	}
+
+	if bucketAccess.Status.AccessGranted {
+		klog.V(5).InfoS("AccessAlreadyGranted",
+			"bucketaccess", bucketAccess.Name,
+			"bucket", bucket.Name,
+		)
+		return nil
+	}
+
+	if bucket.Spec.BucketID == "" {
+		err := errors.New("BucketID cannot be empty")
+		klog.ErrorS(err,
+			"Invalid arguments",
+			"bucket", bucket.Name,
+			"bucketAccess", bucketAccess.Name,
+		)
+		return errors.Wrap(err, "Invalid arguments")
+	}
+
+	req := &cosi.ProvisionerGrantBucketAccessRequest{
+		BucketId:     bucket.Spec.BucketID,
+		AccountName:  bucketAccess.Name,
+		AccessPolicy: bucketAccess.Spec.PolicyActionsConfigMapData,
+	}
+
+	// This needs to be idempotent
+	rsp, err := bal.provisionerClient.ProvisionerGrantBucketAccess(ctx, req)
+	if err != nil {
+		if status.Code(err) != codes.AlreadyExists {
+			klog.ErrorS(err,
+				"Failed to grant access",
+				"bucketAccess", bucketAccess.Name,
+				"bucket", bucket.Name,
+			)
+			return errors.Wrap(err, "failed to grant access")
+		}
+
+	}
+	ns := bal.namespace
+	mintedSecretName := "ba-" + string(bucketAccess.UID)
+	if _, err := bal.Secrets(ns).Get(ctx, mintedSecretName, metav1.GetOptions{}); err != nil {
```

Comment:
Why do Get then Create? Can you call only Create and test for already-exists?

Reply:
That would work too